trino/auth.py

```diff
@@ -251,13 +281,49 @@ def _get_token(self, token_server, response, **kwargs):
 
         raise exceptions.TrinoAuthError("Exceeded max attempts while getting the token")
 
+    def _get_token_from_cache(self, user, host):
+        with self._token_lock:
+            if self._keyring is not None:
+                # keyring is installed, so we can reuse the same token over multiple threads
+                try:
+                    return self._keyring.get_password(
+                        f"{user}@{host}", "token"
+                    )
+                except self._keyring.errors.NoKeyringError:
+                    pass
+
+            # keyring is not installed, so we fall back to a per connection cache
+            return self._token
+
+    def _store_token_to_cache(self, user, host, token):
+        with self._token_lock:
+            if self._keyring is not None:
+                try:
+                    # keyring is installed, so we can store the token for reuse within multiple threads
+                    self._keyring.set_password(
+                        f"{user}@{host}", "token", token
+                    )
+                except self._keyring.errors.NoKeyringError:
+                    print("Warning: although keyring module is installed no backend has been detected, "
+                          "check https://pypi.org/project/keyring/ for more information.")
+
+            # keyring is not installed, so we store the token within the context of the same connection
+            self._token = token
+
+    @staticmethod
+    def _determine_user_host(url, headers) -> Tuple[str, str]:
+        user = headers.get("X-Trino-User")
```

Comment:
Is it used to create a key for saving a password by keyring?

Reply:
Yes, that's it. Now given that the lifetime of a token isn't generally that long, I would think it's probably OK to get rid of the user. Also the keyring is not installed by default, so it means the user opts in on this behavior.

Reply:
What if we use a different user between connections? The token will be overwritten by the newest one?

Reply:
What if we use a different user between connections? The token will be overwritten by the newest one?

Reply:
When you don't use keyring, the token is stored as instance variable within the Authentication class. So it's not reused unless you would reuse that same instance over multiple connections.

With keyring it will just try to reuse the existing token, only after it expires it will prompt again. Now the earlier code was dependent on the user set in the `trino.dbapi.Connection` object, this user is straight from user input. It is not required to set it up for OAuth connections.

The whole point about using keyring is mostly about providing a default user-friendly experience for user ran CLI applications like dbt, I think it would be quite uncommon if they would have multiple users for the same host and also using OAuth. The workaround would be then to not use keyring in this case. Maybe we need to describe this better in the docs?

I also don't think a lot of server applications will use this webbrowser and keyring feature. They can support redirection by adding their own `handle_redirect_auth_url`.

I also would never install keyring by default, it's the users them self that opt in by installing keyring.